FIX: WSL Paste image does not work

Waxime64 opened the pull request:
## Related issues:  
- https://github.com/openai/codex/issues/4683  
- https://github.com/openai/codex/issues/3939  
- https://github.com/openai/codex/issues/2292  
- https://github.com/openai/codex/issues/7528 (After correction https://github.com/openai/codex/pull/3990)

**Area:** `codex-cli` (image handling / clipboard & file uploads)  
**Platforms affected:** WSL (Ubuntu on Windows 10/11). No behavior change on native Linux/macOS/Windows.

## Summary

This PR fixes image pasting and file uploads when running `codex-cli` inside WSL. Previously, image operations failed silently or with permission errors because paths weren't properly mapped between Windows and WSL filesystems.

## Visual Result

<img width="1118" height="798" alt="image" src="https://github.com/user-attachments/assets/14e10bc4-6b71-4d1f-b2a6-52c0a67dd069" />

## Last Rust-Cli

<img width="1175" height="859" alt="image" src="https://github.com/user-attachments/assets/7ef41e29-9118-42c9-903c-7116d21e1751" />

## Root cause

The CLI assumed native Linux/Windows environments and didn't handle the WSL↔Windows boundary:

- Used Linux paths for files that lived on the Windows host
- Missing path normalization between Windows (`C:\...`) and WSL (`/mnt/c/...`)
- Clipboard access failed under WSL

### Why `Ctrl+V` doesn't work in WSL terminals

Most WSL terminal emulators (Windows Terminal, ConEmu, etc.) intercept `Ctrl+V` at the terminal level to paste text from the Windows clipboard. This keypress never reaches the CLI application itself, so our clipboard image handler never gets triggered. Users need `Ctrl+Alt+V`.

## Changes

### WSL detection & path mapping

- Detects WSL by checking `/proc/sys/kernel/osrelease` and the `WSL_INTEROP` env var  
- Maps Windows drive paths to WSL mount paths (`C:\...` → `/mnt/c/...`)

### Clipboard fallback for WSL

- When clipboard access fails under WSL, falls back to PowerShell to extract images from the Windows clipboard  
- Saves to a temp file and maps the path back to WSL

### UI improvements

- Shows `Ctrl+Alt+V` hint on WSL (many terminals intercept plain `Ctrl+V`)  
- Better error messages for unreadable images

## Performance

- Negligible overhead. The fallback adds a single FS copy to a temp file only when needed.  
- Direct streaming remains the default.

## Files changed

- `protocol/src/lib.rs` – Added platform detection module  
- `protocol/src/models.rs` – Added WSL path mapping for local images  
- `protocol/src/platform.rs` – New module with WSL detection utilities  
- `tui/src/bottom_pane/chat_composer.rs` – Added base64 data URL support and WSL path mapping  
- `tui/src/bottom_pane/footer.rs` – WSL-aware keyboard shortcuts  
- `tui/src/clipboard_paste.rs` – PowerShell clipboard fallback

## How to reproduce the original bug (pre-fix)

1. Run `codex-cli` inside WSL2 on Windows.  
2. Paste an image from the Windows clipboard or drag an image from `C:\...` into the terminal.  
3. Observe that the image is not attached (silent failure) or an error is logged; no artifact reaches the tool.

## How to verify the fix

1. Build this branch and run `codex-cli` inside WSL2.  
2. Paste from clipboard and drag from both Windows and WSL paths.  
3. Confirm that the image appears in the tool and the CLI shows a single concise info line (no warning unless fallback was used).

I’m happy to adjust paths, naming, or split helpers into a separate module if you prefer.

## How to try this branch

If you want to try this before it’s merged, you can use my Git branch:

Repository: https://github.com/Waxime64/codex.git  
Branch: `wsl-image-2`

1. Start WSL on your Windows machine.
2. Clone the repository and switch to the branch:
   ```bash
   git clone https://github.com/Waxime64/codex.git
   cd codex
   git checkout wsl-image-2
   # then go into the Rust workspace root, e.g.:
   cd codex-rs
3. Build the TUI binary:
  cargo build -p codex-tui --bin codex-tui --release
4. Install the binary:
   sudo install -m 0755 target/release/codex-tui /usr/local/bin/codex
5. From the project directory where you want to use Codex, start it with:
   cd /path/to/your/project
   /usr/local/bin/codex

On WSL, use CTRL+ALT+V to paste an image from the Windows clipboard into the chat.

Waxime64 committed:
Fix: Image paste/upload does not work under WSL (Windows Subsystem for Linux) new version

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge https://github.com/openai/codex/pull/6046

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 commented:
<img width="1065" height="730" alt="image" src="https://github.com/user-attachments/assets/f752a961-ccee-49fa-bd96-5dfe7648e6fd" />

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Correction merge prompt_caching

Waxime64 committed:
Correction merge prompt_caching

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'openai:main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

adal-idl commented:
I just updated codex to 
codex-cli 0.61.0
windows 11, wsl2.

CTRL+V sitll does nothing

and when I use ALT+CTRL+V
it says
■ Failed to paste image: no image on clipboard: The clipboard contents were not available in the requested format or the
clipboard is empty.


whereas I'm sure there's an image in the clipboard => I can see it with WIN+V
and it also pastes fine in paint.net too
and it also pastes fines here in github.com too
<img width="1001" height="816" alt="image" src="https://github.com/user-attachments/assets/4cbfd132-23ef-45cd-b862-9b34c49292c7" />

<img width="1115" height="628" alt="image" src="https://github.com/user-attachments/assets/4939196f-c320-440f-9284-86c73b6a7da9" />

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'openai:main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
sprint 30

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'wsl-image-2' of https://github.com/Waxime64/codex into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'wsl-image-2' of https://github.com/Waxime64/codex into wsl-image-2

Waxime64 committed:
Revert "sprint 30"

This reverts commit 5683aa16dd5280fa18aaaf150f7c50eba961e692.

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'wsl-image-2' of https://github.com/Waxime64/codex into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'wsl-image-2' of https://github.com/Waxime64/codex into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 commented:
<img width="1249" height="695" alt="image" src="https://github.com/user-attachments/assets/0493820d-e6c1-4182-b968-178fc5de0db7" />

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'wsl-image-2' of https://github.com/Waxime64/codex into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'wsl-image-2' of https://github.com/Waxime64/codex into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'wsl-image-2' of https://github.com/Waxime64/codex into wsl-image-2

wbdb commented:
In both cases, currently and previously, it worked after manually saving (drag and drop). In both cases, it currently does not work when using CTRL + V.  But when I paste using CTRL + V in Word or similar tools, the image is there.

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'wsl-image-2' of https://github.com/Waxime64/codex into wsl-image-2

dylan-hurd-oai commented on codex-rs/tui/src/clipboard_paste.rs:
```diff
@@ -119,19 +121,94 @@ pub fn paste_image_as_png() -> Result<(Vec<u8>, PastedImageInfo), PasteImageErro
+                        ));
+                    }
+                }
+            }
```
Can we pull this logic into a separate function?

Waxime64 replied:
Splitted

Waxime64 committed:
Merge branch 'main' into wsl-image-2

dylan-hurd-oai commented on codex-rs/tui/src/bottom_pane/footer.rs:
```diff
@@ -352,10 +375,20 @@ const SHORTCUTS: &[ShortcutDescriptor] = &[
+            ShortcutBinding {
+                // Use a plain 'v' binding here; overlay_entry will render the
+                // full "ctrl + alt + v" label when state.is_wsl is true.
+                key: key_hint::plain(KeyCode::Char('v')),
```
instead of adding the special case above, can we just add a new `key_hint::ctrl_alt`?

Waxime64 replied:
Yes, done

dylan-hurd-oai reviewed (changes requested):
Excited to fix this problem, but would like to simplify this approach.

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'wsl-image-2' of https://github.com/Waxime64/codex into wsl-image-2

Waxime64 committed:
Revert https://github.com/openai/codex/pull/6046

draft: feat: Better OS Detection, starting with windows #6046

Waxime64 committed:
Code review correction:

Use is_probably_wsl
Separate function: Clipboard
Use  chatwidget.rs
Add key_hint::ctrl_alt

Waxime64 committed:
Format /etc

Waxime64 committed:
Format /etc

Waxime64 committed:
Correction  is_probably_wsl for use correctly /proc/version and not Linux distro name

Waxime64 commented on codex-rs/core/src/environment_context.rs:
```diff
@@ -21,6 +21,14 @@ pub enum NetworkAccess {
+    pub name: String,
+    pub version: String,
+    pub is_likely_windows_subsystem_for_linux: Option<bool>,
+}
```
Remove ref at codex-rs/core + repair is_probably_wsl for use 
/proc/version: Command give Linux version 6.6.87.2-microsoft-standard-WSL2 (root@439a258ad544) (gcc (GCC) 11.2.0, GNU ld (GNU Binutils) 2.34

Waxime64 committed:
Remove old base64 section

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 commented:
I think I forget nothing

Visual Test
![CodexPaste](https://github.com/user-attachments/assets/c9ac9a34-27b3-41f9-a994-62c896a0f296)

Rust-CI
https://github.com/Waxime64/codex/actions/runs/19915273246

@dylan-hurd-oai

Waxime64 committed:
Remove eprintln

dylan-hurd-oai reviewed (commented):
Looking much better! Some final comments

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
chat_composer.rs: Remove image_dimensions Err correction

clipboard_paste.rs: fallback for is_probably_wsl and remove public convert_windows_path_to_wsl

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

Waxime64 committed:
Merge branch 'main' into wsl-image-2

dylan-hurd-oai commented:
Good find - thanks for investigating. Based on your report, it does not appear to be a regression - i.e. calls that failed before will continue to fail. Therefore I think we can proceed with this fix. Can you open up a new github issue for that specific problem so we can track the new PR accordingly?

etraut-openai commented:
Nice! I'm so happy to see this merged!

@Waxime64, thanks again for the contribution. We really appreciate your diligence and persistence on this one.
